### ISSUE TITLE
ChangeFeed: Investigating Compute Gateway AVAD Archival Tree Bounded Caching

### ISSUE THREAD

**Issue by philipthomas-MSFT** [Azure/azure-cosmos-dotnet-v3#4474]
# Description
Investigating Compute Gateway AVAD Archival Tree Bounded Caching for ChangeFeed.

## Problem statement
Currently, Archival caching is unbounded for CF AVAD on Compute Gateway. Investigation is needed to determine if and when bounded caching is necessary.

Tasks

- [ ] Cache must be tied to tenant container (should be evicted when container is deleted)
- [ ] Investigate max size of our cache and implement some limits
- [ ] Publish counters specific to cache information
- [ ] Trim cached data to minimum required set (consider abbreviations, etc.)

**Comment by kirankumarkolli** [comment]
In Tasks, below two are the high priority and MUST

 Cache must be tied to tenant container (should be evicted when container is deleted)
 Publish counters specific to cache information
 
 On when to emit cache full size information
- Size gets impacted by addition/deletes => On add/delete
- Periodic emitter => Gary shared a related code reference

Please share the design of the cache size computation (need to cover the CPU/resurce constraints)

**Comment by kirankumarkolli** [comment]
Is the proposal above final solution? Or current stage? 

Also there is an open fundamental question about "Is JSON serialized size the right representation of in-memory size"? 
Can you please also cover that aspect?

**Comment by philipthomas-MSFT** [comment]
> Is the proposal above final solution? Or current stage?
> 
> Also there is an open fundamental question about "Is JSON serialized size the right representation of in-memory size"? Can you please also cover that aspect?

@kirankumarkolli  I choose this path because I feel that it is probably the best, which is why, in my documentation, I talked briefly about the other options, and I did strike through them.

After checking many sources, there is no accurate means to determine the actual size of an instanced managed object in .NET. Using **Stream.Length** is an approximation of the data contained within the object without the overhead. Memory is usually going to be slightly larger.

If we were dealing with structs, there are things like **Marshall.SizeOf** that we could use, but it is still just an approximation.

**GC.GetTotalMemory(true)** before and after caches are created or removed can give us a number of what was just add or created, but not the actual size of the caches without subtracting afterSize and beforeSize (afterSize - beforeSize). I don't really have a way of asserting this in a test other that the value is greater every time you add to the cache, and lesser every time you remove from the cache. And 0 if the cache is totally cleared. With this approach you get results like this from one of my tests

One cache item, one service endpoint
```
Post MemorySizes.TryAdd, the total memory size of AsyncArchivalTreeCache.Caches: (1496) bytes.
The memory size removed for https://localhost/ with key((7P[NqJ9[AUL=, 1)) was (1496) bytes.
Post MemorySizes.TryRemove, the total memory size of AsyncArchivalTreeCache.Caches: (0) bytes.
```

Two caches, one service endpoint

```
Post MemorySizes.TryAdd, the total memory size of AsyncArchivalTreeCache.Caches: (1736) bytes.
Post MemorySizes.TryAdd, the total memory size of AsyncArchivalTreeCache.Caches: (1752) bytes.
The memory size removed for https://localhost/ with key((#M1u8uTBIXw=, 2)) was (16) bytes.
The memory size removed for https://localhost/ with key((4Z.(u!yQnb>=, 1)) was (1736) bytes.
Post MemorySizes.TryRemove, the total memory size of AsyncArchivalTreeCache.Caches: (0) bytes.
```

Two caches, two service endpoints.
```
Post MemorySizes.TryAdd, the total memory size of AsyncArchivalTreeCache.Caches: (1408) bytes.
Post MemorySizes.TryAdd, the total memory size of AsyncArchivalTreeCache.Caches: (1424) bytes.
The memory size removed for https://localhost/ with key((YHI0O1mybjR=, 1)) was (1408) bytes.
Post MemorySizes.TryRemove, the total memory size of AsyncArchivalTreeCache.Caches: (16) bytes.
The memory size removed for https://localhost2/ with key((x+A]($XTR!j=, 2)) was (16) bytes.
Post MemorySizes.TryRemove, the total memory size of AsyncArchivalTreeCache.Caches: (0) bytes.
```

There is talk that **BenchmarkDotNet** and **ObjectLayoutInspector** has some functions that supposedly does the trick, but not sure this is something we want to introduce.

**Comment by kirankumarkolli** [comment]
GC is for the whole process and there might be other concurrent allocations/de-allocations happening right? Our goal is to find just ArchivalTree approximation for the cache, process memory counters are already captured. 

ObjectLayoutInspector: ArchivalTree is composed of DrainRoute, SplitGraph types, it's a fixed set of types (even if it evolves it will be a bounded set). If for each type how much space is in-memory, then we can compose the total size right? 

Now for each of those types we can pre-compile/model the size and use as a possibility thoughts?

**Comment by philipthomas-MSFT** [comment]
> GC is for the whole process and there might be other concurrent allocations/de-allocations happening right? Our goal is to find just ArchivalTree approximation for the cache, process memory counters are already captured.
> 
> ObjectLayoutInspector: ArchivalTree is composed of DrainRoute, SplitGraph types, it's a fixed set of types (even if it evolves it will be a bounded set). If for each type how much space is in-memory, then we can compose the total size right?
> 
> Now for each of those types we can pre-compile/model the size and use as a possibility thoughts?

@kirankumarkolli 

I agree with you sentiments on GC, which is why I did not wish to go that route but included it in the PR just as a point of reference just like I did for **Stream.Length**.

I do not know much about **ObjectLayoutInspector**. Maybe I am falsely assuming that this was not a viable option bringing in such a dependency in Compute Gateway so I do not fully understand your question or the direction you are going in when you make references to _DrainRoute_, _SplitGraph_ types. If you are suggesting that **ObjectLayoutInspector** is not a deal breaker, then I will look into it more, but my next steps were to start reaching out to others for recommendations, but I think that we are just drawing this out but I am 100% sure you will disagree with whatever stance I take.

The reason why **Stream.Length** wins for me, at least for this iteration, is that it is a consistent number. It goes up. It goes down. It is something we can use to measure. It is not the accuracy of the number. It is what the number represents. But sure, we can keep going down this rabbit hole.
